opm/simulators/utils/ParallelRestart.cpp

```diff
@@ -102,6 +102,156 @@
       unpack(data, buffer, position, comm, std::integral_constant<bool,true>()); \
   }
 
+namespace
+{
+template<template<class, class> class Map, class Type, class Key>
+std::pair<std::set<Type>, std::vector<std::pair<Key, std::vector<int>>>>
+splitDynMap(const Map<Key, Opm::DynamicState<Type>>& map)
+{
+    // we have to pack the unique ptrs separately, and use an index map
+    // to allow reconstructing the appropriate structures.
+    std::vector<std::pair<Key, std::vector<int>>> asMap;
+    std::set<Type> unique;
+    for (const auto& it : map) {
+        for (const auto& w : it.second.data()) {
+            if (w) {
+                unique.insert(w);
+            }
+        }
+    }
+    for (const auto& it : map) {
+        std::vector<int> idxVec;
+        idxVec.reserve(it.second.size()+1);
+        for (const auto& w : it.second.data()) {
+            int distance = std::distance(unique.begin(), unique.find(w));
+            idxVec.push_back(distance);
+        }
+        idxVec.push_back(it.second.initialRange());
+        asMap.push_back(std::make_pair(it.first, idxVec));
+    }
+
+    return std::make_pair(unique, asMap);
+}
+
+template<class Type>
+std::pair<std::set<Type>, std::vector<int>>
+splitDynState(const Opm::DynamicState<Type>& state)
+{
+    std::set<Type> unique;
+    for (const auto& w : state.data()) {
+        if (w) {
+            unique.insert(w);
+        }
+    }
+    std::vector<int> idxVec;
+    idxVec.reserve(state.data().size()+1);
+    for (const auto& w : state.data()) {
+        int distance = std::distance(unique.begin(), unique.find(w));
```

Comment:
This looks a bit fishy. For `OrderedMap` `T` will be  a `shared_ptr`. If state, for whatever reason, stores a `shared_ptr` which is null, then unique will not have entry for it, hence `distance` will be `unique.size()`.

I guess on the receiving side in `reconstructDynMap` will convert whatever is stored at unique[unique.size()] to a pointer.
In most cases this will not be a `nullptr`

Note that changing the order of packing and unpacking, might have influence in this case. Hopefully, this is the bug your are searching for. HTH

Reply:
while indeed this could be an issue, it's not The issue and does not trigger in this case. I repeat it works if i deserialize directly into the variables; the order thing is just so i can do it *after* instancing the Schedule object using the = operator. but nice catch.